File: go/01_基础语法/02_运算符.go
// ============================================================
// Go 基础语法 - 运算符
// 运行方式：go run 02_运算符.go
// ============================================================

package main

import "fmt"

func main() {
	fmt.Println("=== 1. 算术运算符 ===")
	a, b := 10, 3

	fmt.Println("a =", a, ", b =", b)
	fmt.Println("a + b =", a+b)   // 加法
	fmt.Println("a - b =", a-b)   // 减法
	fmt.Println("a * b =", a*b)   // 乘法
	fmt.Println("a / b =", a/b)   // 除法（整数除法）
	fmt.Println("a % b =", a%b)   // 取模（余数）

	// 自增、自减
	a++
	fmt.Println("a++ 后 a =", a)
	b--
	fmt.Println("b-- 后 b =", b)

	fmt.Println("\n=== 2. 关系运算符 ===")
	x, y := 5, 10

	fmt.Println("x =", x, ", y =", y)
	fmt.Println("x == y :", x == y)   // 等于
	fmt.Println("x != y :", x != y)   // 不等于
	fmt.Println("x > y  :", x > y)    // 大于
	fmt.Println("x < y  :", x < y)    // 小于
	fmt.Println("x >= y :", x >= y)   // 大于等于
	fmt.Println("x <= y :", x <= y)   // 小于等于

	fmt.Println("\n=== 3. 逻辑运算符 ===")
	t, f := true, false

	fmt.Println("t =", t, ", f =", f)
	fmt.Println("t && f :", t && f)   // 逻辑与（AND）
	fmt.Println("t || f :", t || f)   // 逻辑或（OR）
	fmt.Println("!t     :", !t)       // 逻辑非（NOT）
	fmt.Println("!f     :", !f)

	// 短路求值
	fmt.Println("\n短路求值示例：")
	n := 0
	if false && (n == 0) {
		fmt.Println("不会执行到这里")
	}
	fmt.Println("短路后 n =", n, "（n 没有被求值）")

	fmt.Println("\n=== 4. 位运算符 ===")
	m, n2 := 60, 13 // 60 = 0011 1100, 13 = 0000 1101

	fmt.Printf("m = %d (二进制: %08b)\n", m, m)
	fmt.Printf("n = %d (二进制: %08b)\n", n2, n2)
	fmt.Printf("m & n  = %d (二进制: %08b)  按位与\n", m&n2, m&n2)
	fmt.Printf("m | n  = %d (二进制: %08b)  按位或\n", m|n2, m|n2)
	fmt.Printf("m ^ n  = %d (二进制: %08b)  按位异或\n", m^n2, m^n2)
	fmt.Printf("m << 2 = %d (二进制: %08b)  左移\n", m<<2, m<<2)
	fmt.Printf("m >> 2 = %d (二进制: %08b)  右移\n", m>>2, m>>2)

	fmt.Println("\n=== 5. 赋值运算符 ===")
	var c int = 10
	fmt.Println("初始 c =", c)

	c += 5
	fmt.Println("c += 5 后 c =", c)

	c -= 3
	fmt.Println("c -= 3 后 c =", c)

	c *= 2
	fmt.Println("c *= 2 后 c =", c)

	c /= 4
	fmt.Println("c /= 4 后 c =", c)

	c %= 3
	fmt.Println("c %= 3 后 c =", c)

	fmt.Println("\n=== 6. 其他运算符 ===")
	d := 42
	p := &d
	fmt.Println("d 的地址:", p)
	fmt.Println("p 指向的值:", *p)

	*p = 100
	fmt.Println("通过指针修改后 d =", d)

	fmt.Println("\n=== 7. 运算符优先级示例 ===")
	result := 5 + 3*2
	fmt.Println("5 + 3 * 2 =", result, "（先乘后加）")

	result = (5 + 3) * 2
	fmt.Println("(5 + 3) * 2 =", result, "（括号优先）")
}

File: rust/01_基础语法/02_运算符.rs
// ============================================================
// Rust 基础语法 - 运算符
// 编译运行：rustc 02_运算符.rs -o 02_运算符 && ./02_运算符
// ============================================================

fn main() {
    println!("=== 1. 算术运算符 ===");
    let a = 10;
    let b = 3;

    println!("a = {}, b = {}", a, b);
    println!("a + b = {}", a + b);
    println!("a - b = {}", a - b);
    println!("a * b = {}", a * b);
    println!("a / b = {}", a / b);
    println!("a % b = {}", a % b);

    // 自增、自减（Rust 没有 ++ 和 --，使用 += 1 或 -= 1）
    let mut x = 5;
    x += 1;
    println!("x += 1 后 x = {}", x);
    x -= 1;
    println!("x -= 1 后 x = {}", x);

    println!("\n=== 2. 关系运算符 ===");
    let x = 5;
    let y = 10;

    println!("x = {}, y = {}", x, y);
    println!("x == y : {}", x == y);
    println!("x != y : {}", x != y);
    println!("x > y  : {}", x > y);
    println!("x < y  : {}", x < y);
    println!("x >= y : {}", x >= y);
    println!("x <= y : {}", x <= y);

    println!("\n=== 3. 逻辑运算符 ===");
    let t = true;
    let f = false;

    println!("t = {}, f = {}", t, f);
    println!("t && f : {}", t && f);
    println!("t || f : {}", t || f);
    println!("!t     : {}", !t);
    println!("!f     : {}", !f);

    // 短路求值
    println!("\n短路求值示例：");
    let mut n = 0;
    if false && { n += 1; true } {
        println!("不会执行到这里");
    }
    println!("短路后 n = {}（右边没有被求值）", n);

    println!("\n=== 4. 位运算符 ===");
    let m: u8 = 0b0011_1100; // 60
    let n: u8 = 0b0000_1101; // 13

    println!("m = {:08b} ({})", m, m);
    println!("n = {:08b} ({})", n, n);
    println!("m & n  = {:08b} ({})  按位与", m & n, m & n);
    println!("m | n  = {:08b} ({})  按位或", m | n, m | n);
    println!("m ^ n  = {:08b} ({})  按位异或", m ^ n, m ^ n);
    println!("!m     = {:08b} ({})  按位取反", !m, !m);
    println!("m << 2 = {:08b} ({})  左移", m << 2, m << 2);
    println!("m >> 2 = {:08b} ({})  右移", m >> 2, m >> 2);

    println!("\n=== 5. 赋值运算符 ===");
    let mut c = 10;
    println!("初始 c = {}", c);

    c += 5;
    println!("c += 5 后 c = {}", c);

    c -= 3;
    println!("c -= 3 后 c = {}", c);

    c *= 2;
    println!("c *= 2 后 c = {}", c);

    c /= 4;
    println!("c /= 4 后 c = {}", c);

    c %= 3;
    println!("c %= 3 后 c = {}", c);

    println!("\n=== 6. 其他运算符 ===");

    // 引用和解引用
    let d = 42;
    let p = &d;
    println!("d 的引用: {}", p);
    println!("解引用 p: {}", *p);

    // 范围运算符
    println!("\n--- 范围运算符 ---");
    println!("1..5 :");
    for i in 1..5 {
        print!("{} ", i);
    }
    println!();

    println!("1..=5 :");
    for i in 1..=5 {
        print!("{} ", i);
    }
    println!();

    println!("\n=== 7. 运算符优先级示例 ===");
    let result = 5 + 3 * 2;
    println!("5 + 3 * 2 = {}（先乘后加）", result);

    let result = (5 + 3) * 2;
    println!("(5 + 3) * 2 = {}（括号优先）", result);
}
